### ISSUE TITLE
FN collection_is_never_read: grabbing unit value of push()

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#10488]
### Summary

,

### Lint Name

collection_is_never_read

### Reproducer

I tried this code:

```rust
#![warn(clippy::collection_is_never_read)]

pub fn main() {
    let mut x: Vec<_> = vec![1, 2, 3];
    x.push(4);
    // the println() is read as a use, but the .push() returns () so we still never actually use the resulting vec
    println!("{:?}", x.push(5));
}
```

I expected to see this happen:
Lint trigggers

Instead, this happened:
Linnt doesn't trigger

### Version

```text
rustc 1.70.0-nightly (8a73f50d8 2023-03-11)
binary: rustc
commit-hash: 8a73f50d875840b8077b8ec080fa41881d7ce40d
commit-date: 2023-03-11
host: x86_64-unknown-linux-gnu
release: 1.70.0-nightly
LLVM version: 15.0.7
```

**Comment by schubart** [comment]
This is known / intended behaviour. There's a test for it:

```rust
#[allow(clippy::let_unit_value)]
fn fake_read() {
    let mut x = vec![1, 2, 3]; // Ok
    x.reverse();
    // `collection_is_never_read` gets fooled, but other lints should catch this.
    let _: () = x.clear();
}
```

Aren't there other lints that catch that `println!("{:?}", x.push(5));` is pointless? Are there valid use cases for such code?

I wonder if instead of special-casing the unit type in `collection_is_never_read`, there should be a new lint that catches this kind of "usage" of unit type values.
